File: ueueo-framework/ueueo-domain-driven-desgin/src/main/java/com/ueueo/ddd/domain/entities/events/EntityEventData.java
package com.ueueo.ddd.domain.entities.events;

import com.ueueo.ID;
import com.ueueo.eventbus.IEventDataMayHaveTenantId;
import com.ueueo.eventbus.IEventDataWithInheritableGenericArgument;
import com.ueueo.multitenancy.IMultiTenant;
import lombok.Getter;

/**
 * Used to pass data for an event that is related to with an <see cref="IEntity"/> object.
 *
 * @author Lee
 * @date 2022-05-20 14:20
 */
public class EntityEventData<TEntity> implements IEventDataWithInheritableGenericArgument, IEventDataMayHaveTenantId {
    @Getter
    private TEntity entity;

    public EntityEventData(TEntity entity) {
        this.entity = entity;
    }

    @Override
    public Object[] getConstructorArgs() {
        return new Object[]{entity};
    }

    @Override
    public boolean isMultiTenant() {
        return entity instanceof IMultiTenant;
    }

    @Override
    public ID getTenantId() {
        if (entity instanceof IMultiTenant) {
            return ((IMultiTenant) entity).getTenantId();
        }
        return null;
    }
}

File: ueueo-framework/ueueo-domain-driven-desgin/src/main/java/com/ueueo/ddd/domain/entities/events/EntityEventData.cs
using System;
using Volo.Abp.EventBus;
using Volo.Abp.MultiTenancy;

namespace Volo.Abp.Domain.Entities.Events;

/**
 * Used to pass data for an event that is related to with an <see cref="IEntity"/> object.
*
 * <typeparam name="TEntity">Entity type</typeparam>
     */
[Serializable]
public class EntityEventData<TEntity> : IEventDataWithInheritableGenericArgument, IEventDataMayHaveTenantId
{
    /**
     * Related entity with this event.
    */
    public TEntity Entity;//  { get; }

    /**
     * Constructor.
    *
     * <param name="entity">Related entity with this event</param>
     */
    public EntityEventData(TEntity entity)
    {
        Entity = entity;
    }

    public   Object[] GetConstructorArgs()
    {
        return new Object[] { Entity };
    }

    public   boolean IsMultiTenant(out ID tenantId)
    {
        if (Entity is IMultiTenant multiTenantEntity)
        {
            tenantId = multiTenantEntity.TenantId;
            return true;
        }

        tenantId = null;
        return false;
    }
}
